apps/cnspec/cmd/serve.go

```diff
@@ -0,0 +1,192 @@
+	},
+	Run: func(cmd *cobra.Command, args []string) {
+		prof.InitProfiler()
+
+		// prevent colors on windows
+		viper.Set("color", "none")
+
+		// check if an inventory file exists
+		inventoryFilePath, ok := config.InventoryPath(viper.ConfigFileUsed())
+		if ok {
+			viper.Set("inventory", inventoryFilePath)
+		}
+
+		// determine the scan config from pipe or args
+		conf, err := getServeConfig()
+		if err != nil {
+			log.Error().Err(err).Msg("could not load configuration")
+			// we return the specific error code to prevent systemd from restarting
+			os.Exit(cnquery_cmd.ConfigurationErrorCode)
+		}
+
+		ctx := cnquery.SetFeatures(context.Background(), cnquery.DefaultFeatures)
+
+		if conf != nil && conf.UpstreamConfig != nil {
+			hc := backgroundjob.NewHealthPinger(ctx, conf.UpstreamConfig.ApiEndpoint, 5*time.Minute)
+			hc.Start()
+			defer hc.Stop()
+		}
+
+		bj, err := backgroundjob.New()
+		if err != nil {
+			log.Fatal().Err(err).Msg("could not start background listener")
+		}
+
+		bj.Run(func() error {
+			// TODO: check in every 5 min via timer, init time in Background job
+			result, err := RunScan(conf, scan.DisableProgressBar())
+			if err != nil {
+				log.Error().Err(err).Msg("could not successfully complete scan")
+			}
+
+			// log errors
+			if result != nil && result.GetErrors() != nil && len(result.GetErrors()) > 0 {
+				assetErrors := result.GetErrors()
+				for a, err := range assetErrors {
+					log.Error().Err(errors.New(err)).Str("asset", a).Msg("could not connect to asset")
+				}
+			}
+			return nil
+		})
+	},
+}
+
+func getServeConfig() (*scanConfig, error) {
+	opts, optsErr := cnspec_config.ReadConfig()
+	if optsErr != nil {
+		log.Fatal().Err(optsErr).Msg("could not load configuration")
+	}
+	config.DisplayUsedConfig()
+
+	logClientInfo(opts.SpaceMrn, opts.AgentMrn, opts.ServiceAccountMrn)
+
+	// display activated features
+	if len(opts.Features) > 0 {
+		log.Info().Strs("features", opts.Features).Msg("user activated features")
+	}
+
+	conf := scanConfig{
+		Features:   opts.GetFeatures(),
+		DoRecord:   viper.GetBool("record"),
+		ReportType: scan.ReportType_ERROR,
+		Output:     "",
+	}
+
+	// detect CI/CD runs and read labels from runtime and apply them to all assets in the inventory
+	runtimeEnv := execruntime.Detect()
+	if opts.AutoDetectCICDCategory && runtimeEnv.IsAutomatedEnv() || opts.Category == "cicd" {
+		log.Info().Msg("detected ci-cd environment")
+		// NOTE: we only apply those runtime environment labels for CI/CD runs to ensure other assets from the
+		// inventory are not touched, we may consider to add the data to the flagAsset
+		if runtimeEnv != nil {
+			runtimeLabels := runtimeEnv.Labels()
+			conf.Inventory.ApplyLabels(runtimeLabels)
+		}
+		conf.Inventory.ApplyCategory(asset.AssetCategory_CATEGORY_CICD)
+	}
+
+	serviceAccount := opts.GetServiceCredential()
+	if serviceAccount != nil {
+		certAuth, err := upstream.NewServiceAccountRangerPlugin(serviceAccount)
+		if serviceAccount == nil {
```

Comment:
Is this intended?

I would have expected something like `certAuth == nil` or `err != nil`.

Reply:
yes that is plain wrong.